### PR TITLE
OSDOCS#12209: Sigstore signature verification for core OCP images

### PR DESCRIPTION
**Pull request by skopacz1**
[OSDOCS-11885](https://issues.redhat.com/browse/OSDOCS-11885)

Version(s): 4.17+

This PR adds 1. general content about how OCP validates the integrity and authenticity of container images and 2. specific details about how the 4.17 TP ContainerImagePolicy resource plays a part in the authenticity validation.

QE review:
- [ ] QE has approved this change.

Preview: [Container security and validation](https://81985--ocpdocs-pr.netlify.app/openshift-enterprise/latest/security/container_security/security-understanding.html#security-understanding-container-validation_security-understanding)

### CODE REVIEW

File: modules/security-understanding-container-validation.adoc [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+The previous example resources define two policies, both of which apply to `test0.com` pullspecs, and only one of which applies to `test1.com` pullspecs.
+When image integrity is validated at pull time for images whose pullspecs have configured policies with matching `scopes` values, CRI-O will retrieve any Sigstore signature images associated with the image and check them against the configured policies.
+If signature verification fails for any configured policy, the associated pod will be rejected with <missing_info> conditions in Pod status and events in the associated namespace.
+////
+FIXME: Should "associated pod" be "associated container"?
+FIXME: specific conditions needed for Pod status
+////
+
+[id="security-understanding-trusting-install_{context}"]
+== Trusting an {product-title} installation
```
Comment (mtrmac) [inline modules/security-understanding-container-validation.adoc:109]: Do you also want to have a section about trusting updates, describing the Cincinnati stream?

It might not be relevant for this article, I don’t know.

File: modules/security-understanding-container-validation.adoc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  policy:
+    rootOfTrust:
+      policyType: PublicKey
+      publicKey:
+        keyData: dGVzdC1rZXktZGF0YQ==
+----
+<1> This policy for `test0.com` and the policy from `mypolicy-0` will be appended together.
+
+The previous example resources define two policies, both of which apply to `test0.com` pullspecs, and only one of which applies to `test1.com` pullspecs.
+When image integrity is validated at pull time for images whose pullspecs have configured policies with matching `scopes` values, CRI-O will retrieve any Sigstore signature images associated with the image and check them against the configured policies.
```
Comment (mtrmac) [inline modules/security-understanding-container-validation.adoc:101]: This should probably be a bit more precise about what “matching” means … and I should double-check.
In the underlying c/image, only the _closest_ matching scope is used; I’m not immediately sure wether `{Cluster,}ImagePolicy` also uses the “closest” scope semantics or whether the wider scopes are “inherited” down into the narrower ones.

File: modules/security-understanding-container-validation.adoc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        prefix: test-remap-prefix
+        signedPrefix: test-remap-signed-prefix
+
+----
+<1> Specifies OIDC issuer and the email of the Fulcio authentication configuration.
+
+.Example `ClusterImagePolicy` resource 2
+[source,yaml]
+----
+kind: ClusterImagePolicy
```
Comment (QiWang19) [inline modules/security-understanding-container-validation.adoc:85]:
```suggestion
apiVersion: config.openshift.io/v1alpha1
kind: ClusterImagePolicy
```

File: modules/security-understanding-container-validation.adoc [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  policy:
+    rootOfTrust:
+      policyType: FulcioCAWithRekor
+      fulcioCAWithRekor:
+        fulcioCAData: dGVzdC1jYS1kYXRhLWRhdGE=
+        fulcioSubject: <example_input_here> <1>
+        rekorKeyData: dGVzdC1yZWtvci1rZXktZGF0YQ==
+    fulcioSubject:
+      oidcIssuer: https://OIDC.example.com
+      signedEmail: test-user@example.com
```
Comment (QiWang19) [inline modules/security-understanding-container-validation.adoc:72]:
```suggestion
      fulcioCAWithRekor:
        fulcioCAData: dGVzdC1jYS1kYXRhLWRhdGE=
        rekorKeyData: dGVzdC1yZWtvci1rZXktZGF0YQ==
        fulcioSubject:
          oidcIssuer: https://OIDC.example.com
          signedEmail: test-user@example.com
```

File: modules/security-understanding-container-validation.adoc [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+. The image is pulled by specifying the SHA256 image digest, for example `quay.io/openshift-release-dev/ocp-release@sha256:ea7ac3ad42169b39fce07e5e53403a028644810bee9a212e7456074894df40f3`.
+
+. The image signature is also being validated. See “Image authenticity validation” for more information.
+====
+
+[id="security-understanding-image-authenticity_{context}"]
+== Image authenticity validation
+
+In addition to validating the integrity of images, {product-title} clusters also validate their authenticity by certifying that the images were approved by a trusted source, in this case Red{nbsp}Hat.
```
[the remark this answers is not in the document]
Comment (wking) [reply modules/security-understanding-container-validation.adoc:46]: > In the context of OpenShift installs/updates, right now, this version binding IIRC not used...

Correct.  But the only images Red Hat is currently signing under `quay.io/openshift-release-dev/ocp-release` are OCP release images.  And when a version string is part of the update request, divergence between the requested version and the actual release image version is fatal since 2020's openshift/cluster-version-operator#431.  So the "is the requested image the OCP release image you're hoping for?" are happening today, just not at the pullspec-asserting-signature level.

File: modules/security-understanding-container-validation.adoc (hunk 2 above)
Comment (wking) [reply modules/security-understanding-container-validation.adoc:101]: [ClusterImagePolicy API Godocs][1] include:

> If multiple scopes match a given image, only the policy requirements for the most specific scope apply. The policy requirements for more general scopes are ignored.

[1]: https://github.com/openshift/api/blob/07bccfd9266faa9565b51fad64e7a773bfb0f8a6/config/v1alpha1/types_cluster_image_policy.go#L41C5-L42C5

File: modules/security-understanding-container-validation.adoc [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      publicKey:
+        keyData: dGVzdC1rZXktZGF0YQ==
+----
+<1> This policy for `test0.com` and the policy from `mypolicy-0` will be appended together.
+
+The previous example resources define two policies, both of which apply to `test0.com` pullspecs, and only one of which applies to `test1.com` pullspecs.
+When image integrity is validated at pull time for images whose pullspecs have configured policies with matching `scopes` values, CRI-O will retrieve any Sigstore signature images associated with the image and check them against the configured policies.
+If signature verification fails for any configured policy, the associated pod will be rejected with <missing_info> conditions in Pod status and events in the associated namespace.
+////
+FIXME: Should "associated pod" be "associated container"?
```
[the remark this answers is not in the document]
Comment (wking) [reply modules/security-understanding-container-validation.adoc:104]: Pod is the Kube resource, so it is the Pod's `status`.  Within that Pod's status, there are `containerStatuses` and `initContainerStatus` and `ephemeralContainerStatuses`, and the `SignatureValidationFailed` and `ImagePullBackOff` reason/messaging will be tucked down in there for impacted container(s).

[the remark this answers is not in the document]
Comment (mtrmac) [reply modules/security-understanding-container-validation.adoc:104]: Thanks, I should have known these things.

File: modules/security-understanding-container-validation.adoc [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      oidcIssuer: https://OIDC.example.com
+      signedEmail: test-user@example.com
+    signedIdentity:
+      matchPolicy: RemapIdentity
+      remapIdentity:
+        prefix: test-remap-prefix
+        signedPrefix: test-remap-signed-prefix
+
+----
+<1> Specifies OIDC issuer and the email of the Fulcio authentication configuration.
```
Comment (saschagrunert) [inline modules/security-understanding-container-validation.adoc:80]: I think we should also explain the other fields and their possible values.

File: modules/security-understanding-container-validation.adoc [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+== Image integrity validation at pull time
+
+When images are downloaded, or pulled, from a registry, an {product-title} cluster performs the following actions to ensure the integrity of each image:
+
+. The cluster downloads an image manifest and calculates its digest.
+
+. The cluster compares the calculated digest of the image manifest with the known digest.
+If the digest values do not match, the image is rejected without further processing.
+
+. The cluster downloads an image config file, which contains metadata such as SHA256 digests for each part of the container image, and calculates the digest of the image config file.
```
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:22]: I think it would be fair to trim out unnecessary details out. I think I added that part in to contextualize what an image config file even is, so it wouldn't just say "there's an image config file and that's part of the validation".

How about I change this line to "The cluster downloads an image config file and calculates its digest." to keep it simple?

File: modules/security-understanding-container-validation.adoc [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+. For each file that comprises the container image, the cluster downloads the file and calculates its digest.
+
+. The cluster compares the calculated digest of each file that makes up the image with the corresponding digest that is provided in the image manifest.
+If the digest values do not match, the image is rejected.
+
+Once the cluster completes all of the previous steps, the integrity of the image has been validated.
+
+[NOTE]
+====
+The image manifest is validated only when the manifest digest is known, which is true in the following cases:
```
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:36]: Is there any additional content you'd suggest to add this info in?

At the end of this note, should I add something like "Even if the image manifest is not validated in these cases, the image config and layers are still checked for integrity"?

File: modules/security-understanding-container-validation.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:46]: The original word used was "published", though @wking suggested to change it to "approved" with the following reasoning given:

> The signer and publisher can be different, e.g. a customer's ProdSec team might sign some subset of Red Hat release images as acceptable, and then those releases would have both Red Hat and customer-ProdSec signatures

So I'm not sure where to proceed if "approved" is too ambiguous. Maybe some rewording of that paragraph to convey the point about "was produced by a RH build system and not modified since"? Maybe something like this:
```suggestion
In addition to validating the integrity of images, {product-title} clusters also validate their authenticity by certifying that the images have not been modified since being produced by a trusted source, in this case Red{nbsp}Hat.
```

File: modules/security-understanding-container-validation.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply modules/security-understanding-container-validation.adoc:22]: Nothing else in the article specifically discusses the image config, so I’m not sure it’s necessary to contextualize it.

… but I wouldn’t know what customers are asking about.

File: modules/security-understanding-container-validation.adoc (hunk 7 above)
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:80]: Sure, is there any content that describes these parameters, or could anyone provide me such information?

File: modules/security-understanding-container-validation.adoc (hunk 3 above)
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:85]: Sounds good, does this need to be added to the other example too?

File: modules/security-understanding-container-validation.adoc (hunk 2 above)
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:101]: What would the wording change look like in light of this feedback? Or do I need to add another sentence to clarify what matching means?

File: modules/security-understanding-container-validation.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:104]: Per conversation with @wking I'll reword this to something like the following:

> If signature verification fails for any configured policy, the associated image pull will be rejected.  Events will be created in the associated namespace explaining the verification failure, and the Pod `status` for containers consuming that image will set a relevant `state`.

But let me know if there's any further feedback for that new wording

File: modules/security-understanding-container-validation.adoc (hunk 1 above)
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:109]: I can add that section from [this blog](https://www.redhat.com/en/blog/certifiable-software-distribution-process-for-openshift), which I'm using as a reference for this overall PR, if you think it would be valuable to add. Or if it's something that can wait until a v2 of this doc, then I'm okay with that, too.

File: modules/security-understanding-container-validation.adoc (hunk 7 above)
Comment (wking) [reply modules/security-understanding-container-validation.adoc:80]: There are Godocs like [these][1].  But I'd expect those to get rendered under [here][2].  Or maybe we only doc GA-cluster APIs there?  For these longer-form docs, I'd expect something that weaves some user stories about how the knobs all come together to deliver a valuable target, with notes to allow customers to riff off that target, and I'm not aware of anything existing today in that space.  For example [this enhancement user story][3] is very thin, and the policy YAML quoted later in the enhancement doesn't have associated text explaining the environment it fits into and how the configuration plugs into that environment.

I agree that these kind of longer stories would be very useful.  I think node team can add them as follow-up improvements on an initial pass in this pull, because y'all are the experts on Sigstore and the sorts of environments customers are likely to set up and want to plug their policies into.

[1]: https://github.com/openshift/api/blob/8f6ded478e8ab589ade8749fb3b70a0fb90a0e60/config/v1alpha1/types_image_policy.go#L181-L195
[2]: https://docs.openshift.com/container-platform/4.16/rest_api/config_apis/config-apis-index.html
[3]: https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-ClusterImagePolicy-and-ImagePolicy-for-Signature-Verification.md#user-stories

File: modules/security-understanding-container-validation.adoc (hunk 3 above)
Comment (wking) [reply modules/security-understanding-container-validation.adoc:85]: yeah, all the ClusterImagePolicy YAML examples will need this `apiVersion` line

File: modules/security-understanding-container-validation.adoc (hunk 2 above)
Comment (wking) [reply modules/security-understanding-container-validation.adoc:101]: I'd just inline the Godocs sentence after the one we're commenting on here, so folks here about how "maching `scopes`..." matter, and then the next sentence gives more precision about what qualifies as "matching".

File: modules/security-understanding-container-validation.adoc (hunk 1 above)
Comment (wking) [reply modules/security-understanding-container-validation.adoc:109]: I think we can exclude Cincinnati/OCP updates to ship a v1 of this content, and then we can think through if/how to add something in that space in follow-up work.

File: modules/security-understanding-container-validation.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (wking) [reply modules/security-understanding-container-validation.adoc:46]: I'm going to step back and suggest that:

> In addition to validating the integrity of images, {product-title} clusters also validate their authenticity by certifying that the images were approved by a trusted source, in this case Red{nbsp}Hat.
> This validation process includes checking the digital signature of the images.
>
> Cluster administrators can use `ClusterImagePolicy` resources to declare image policy for the cluster.

get reworded to something like:

> Cluster administrators can use `ClusterImagePolicy` resources to declare image policy for the cluster, requiring signature validation and other policy checks in addition to image integrity.

Although other wording in the current pull is a bit mushy on whether "integrity" includes policy, or is limited to the internally-consistent digest-matching and excluding higher-level policy like sufficient signatures.

[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:46]: Okay, I'll go with your suggested wording for now, and if need be it can be further refined here or in a v2 of this doc.

File: modules/security-understanding-container-validation.adoc (hunk 7 above)
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:80]: Would explaining more of these parameters here honestly be useful in the context of this particular doc? I feel like the purpose of this doc is to mostly just show that ClusterImagePolicy exists and that it now plays a role in internal core image validation.

To me, it seems like if we really wanted to show users how they could use this feature for their own consumption, that would belong in some different doc, perhaps some procedure, and that this other doc would be a better location for any sort of comprehensive reference info.

File: modules/security-understanding-container-validation.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:36]: Added my suggested sentence for now

File: modules/security-understanding-container-validation.adoc (hunk 2 above)
Comment (skopacz1) [reply modules/security-understanding-container-validation.adoc:103]: Okay, I added the text snippet you provided

File: modules/security-understanding-container-validation.adoc (hunk 7 above)
Comment (mtrmac) [reply modules/security-understanding-container-validation.adoc]: I generally think that there only needs to be one place for full reference documentation, and this article does not look like the canonical place for that. Linking to the reference wouldn’t hurt.

File: modules/security-understanding-container-validation.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply modules/security-understanding-container-validation.adoc:36]: “for integrity against the [untrusted] manifest”? I’m very unsure about the “untrusted” word — it’s correct in the security jargon, but it might carry the connotation of “specifically known to be dangerous”.

My general opinion is that none of the details of configs/layers need to be included, and then we don’t need to discuss this either; but if they are discussed, like that.

File: modules/security-understanding-container-validation.adoc [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+        fulcioCAData: dGVzdC1jYS1kYXRhLWRhdGE=
+        rekorKeyData: dGVzdC1yZWtvci1rZXktZGF0YQ==
+        fulcioSubject: <1>
+          oidcIssuer: https://OIDC.example.com
+          signedEmail: test-user@example.com
+    signedIdentity:
+      matchPolicy: RemapIdentity
+      remapIdentity:
+        prefix: test-remap-prefix
+        signedPrefix: test-remap-signed-prefix
```
Comment (mtrmac) [inline modules/security-understanding-container-validation.adoc:96]: (For the record, these values are invalid will either be rejected, or won’t do anything — real prefixes contain at least a host name with at least one dot, or a host:port. But then again, the `…Data` fields don’t contain really realistic inputs either.)

Comment (mtrmac) [reply modules/security-understanding-container-validation.adoc:96]: I think that _if we care_ about the example values, it would be better to make this all realistic. `scopes: … registry.internal`; `prefix: registry.internal/vendorA/productA`; `signedPrefix: vendorA.example/productA` or something (and maybe the same `vendorA.example` in the Fulcio conditions).

But, also, it seems to me that details of these CRs are really not the primary focus of the article.

File: modules/security-understanding-container-validation.adoc [hunk 11]
```diff
@@ -0,0 +1,159 @@
+// Module included in the following assemblies:
+//
+// * security/container_security/security-understanding.adoc
+
+:_mod-docs-content-type: CONCEPT
+[id="security-understanding-container-validation_{context}"]
+= Container security and validation
+
+{product-title} is released as a set of container images that comprise the different functions of the software.
+These images must be validated in order to ensure that they have not been tampered with by an external party, and to ensure that they are the intended images provided by Red{nbsp}Hat.
```
Comment (wking) [inline modules/security-understanding-container-validation.adoc:10]: Maybe we can rephrase to highlight "customer admins can use ClusterImagePolicy for their own policy too!" with something like:

> {product-title} runs containers from containers images.
> These images may be validated in order to ensure that they have not been tampered with by an external party.

Then the following sections are:

* _Image integrity validation at pull time_, which can be [dropped][1] or tweaked.
* _Image authenticity validation_, introducing ClusterImagePolicy as a generic tool.
* _Trusting an {product-title} installation_, which gets into Red Hat images for the first time.
  * Unfortunately for continuity, we have to take a bit of a detour into bootstrapping/installer trust, which is complicated.
  * _Trusting a connected installation_, still talking about client/installer trust.
  * _Trusting a disconnected installation_, still talking about client/installer trust.

Hrm, and seems like we want an additional section like _Trusting an {product-title} post-install_, that points out the trusted installer will roll out a trusted release image with [the `openshift` ClusterImagePolicy][2] (for TechPreview clusters), and that that pulls in the ClusterImagePolicy security discussed under _Image authenticity validation_ for images in the `quay.io/openshift-release-dev/ocp-release` scope.  That prevents folks from running an `quay.io/openshift-release-dev/ocp-release` image that hasn't been signed by the trusted Red Hat key used to sign OCP release images, e.g. in the event that quay.io or a customer-configured mirror registry is compromised.

Or maybe that only matters when we start to talk about updates, and [we're puning on that for now][3], because really, you'd need a compromised mirror and an admin tricked into updating to a malicious digest or a by-tag release image reference.

[1]: https://github.com/openshift/openshift-docs/pull/81985#discussion_r1767389420
[2]: https://github.com/openshift/enhancements/blob/master/enhancements/security/openshift-image-policy.md#cluster-update-keys
[3]: https://github.com/openshift/openshift-docs/pull/81985#discussion_r1767085205

Comment (mtrmac) [reply modules/security-understanding-container-validation.adoc:10]: Where I’m a little lost is that the goal, per the title _of the PR_ (not of the document), is to talk about OCP product images … which don’t currently use sigstore / ``ClusterImagePolicy` IIRC (just in a tech preview). So if that is the goal, why bring `ClusterImagePolicy` into the picture at all, _now_? It would certainly be relevant in a future release, but I don’t think we are there yet.

OTOH the title of the document (as opposed to the PR) talks about container “security and validation”, i.e. not restricted to core OCP, and in that case `ClusterImagePolicy` is certainly the most relevant part, and the OCP trust bootstrapping a side issue.

File: modules/security-understanding-container-validation.adoc [hunk 12]
```diff
@@ -0,0 +1,159 @@
+// Module included in the following assemblies:
+//
+// * security/container_security/security-understanding.adoc
+
+:_mod-docs-content-type: CONCEPT
+[id="security-understanding-container-validation_{context}"]
+= Container security and validation
```
Comment (mtrmac) [inline modules/security-understanding-container-validation.adoc:7]: Pedantic: “container security” is much larger than the scope of this document.

“Container image autenticity?” “Preventing supply chain attacks on container images”?

File: modules/security-understanding-container-validation.adoc (hunk 11 above)
Comment (wking) [reply modules/security-understanding-container-validation.adoc:10]: OCP trust bootstrapping is more of an installer issue, and I'm fine with them leading there, and doc'ing the long-GA processes around that.

ClusterImagePolicy is more of a node issue, and I'm fine with the node folks leading on that.  There was some work associated with the 4.16 tech-preview of the type in #76613, #77203, #78396.

I'm here because 4.17 is releasing the tech-preview `openshift` ClusterImagePolicy roughly as laid out in [this enhancement][1].  There's a release note landed today via #81997, but it doesn't cover mechanics, and I'd expect customer security folks to have some questions about mechanics.  But maybe they don't?  I'm fine letting this float until there's some clarity around the kinds of security questions customers ask, and if anyone is particularly interested in if/how to manage installer trust or have customers create their own ClusterImagePolicies.

[1]: https://github.com/openshift/enhancements/blob/master/enhancements/security/openshift-image-policy.md

**Comment by kalexand-rh** [comment]
The `branch/enterprise-4.18` label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.17. And any PR going into `main` must also target the latest version branch (enterprise-4.18).

If the update in your PR does NOT apply to version 4.18 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

**Comment by QiWang19** [comment]
> So if that is the goal, why bring ClusterImagePolicy into the picture at all, now? It would certainly be relevant in a future release, but I don’t think we are there yet.

FYI the https://github.com/openshift/openshift-docs/pull/86087 this ongoing PR add the document for ClusterImagePolicy and ImagePolicy

**Comment by wking** [comment]
#86087 is up now with some more-Node-side noodling in this space, although it currently doesn't spend much time explaining how the `openshift` ClusterImagePolicy drives secure OCP updates.

**Comment by bergerhoffer** [comment]
The `branch/enterprise-4.19` label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.18. And any PR going into `main` must also target the latest version branch (enterprise-4.19).

If the update in your PR does NOT apply to version 4.19 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

**Comment by skopacz1** [comment]
Can't guarantee I'll actually get to this with my other assignments, so for now I'm closing this PR. Can reopen if the direction of this work is now clear, or if another writer wants to take on this work then they can feel free to take this off my hands.
